### PR TITLE
Search: Document event context, groups and orders.

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Comment by erikjohnston** [comment]
And also document profile information and current state options.

**Comment by NegativeMjark** [comment]
LGTM

File: specification/modules/search.rst [hunk 1]
```diff
@@ -40,6 +40,41 @@ The value of ``count`` may not match the number of results. For example due to
...
+The client can request that the results are returned along with grouping
+information, e.g. grouped by ``room_id``. In this case the response will
+contain a group entry for each distinct value of ``room_id``. Each group entry
+contains at least a list of the ``event_ids`` that are in that group, as well
+as potentially other metadata about the group.
+
+Pagination
+----------
+
+The server may return a ``next_batch`` key at various places in the response.
```
Comment (ara4n) [inline specification/modules/search.rst]: "Various places" is way too ambiguous :(

File: specification/modules/search.rst [hunk 2]
```diff
@@ -40,6 +40,41 @@ The value of ``count`` may not match the number of results. For example due to
...
+The default ordering is ``rank``.
+
+Groups
+------
+
+The client can request that the results are returned along with grouping
+information, e.g. grouped by ``room_id``. In this case the response will
+contain a group entry for each distinct value of ``room_id``. Each group entry
+contains at least a list of the ``event_ids`` that are in that group, as well
+as potentially other metadata about the group.
```
Comment (ara4n) [inline specification/modules/search.rst]: If this is the only way we can get the right 'recents' ordering, we need to explicitly say so.  However, I still think the results should be returned in an array for simplicity unless we're doing particularly exotic groupings (and I can't think of any where groups overlap?)

File: specification/modules/search.rst (hunk 1 above)
Comment (erikjohnston) [inline specification/modules/search.rst]: I added a note further down about exactly where ``next_batch`` tokens can appear.

**Comment by erikjohnston** [comment]
I've changed the format to have results returned as an ordered list.

E.g.:

```json
{
  "search_categories": {
    "room_events": {
      "count": 24,
      "next_batch": "5FdgFsd234dfgsdfFD",
      "results": [
        {
          "rank": 0.00424866,
          "result": {
            "age": 526228296,
            "content": {
              "body": "Test content martians and men",
              "msgtype": "m.text"
            },
            "event_id": "$144429830826TWwbB:localhost",
            "origin_server_ts": 1444298308034,
            "room_id": "!qPewotXpIctQySfjSy:localhost",
            "type": "m.room.message",
            "user_id": "@test:localhost"
          }
        }
      ]
    }
  }
}

```

**Comment by ara4n** [comment]
what is the decision on leaving in or discarding the results array of event-ids in groupings?

**Comment by erikjohnston** [comment]
> what is the decision on leaving in or discarding the results array of event-ids in groupings?

I haven't really come to a conclusion yet. We ideally do what some structure about so that people can ask for more results for a particular group.

File: api/client-server/search.yaml [hunk 3]
```diff
@@ -68,6 +76,65 @@ paths:
...
+                        type: boolean
+                        title: Include current state
+                        description: |-
+                          Requests the server return the current state for
+                          each room returned.
+                      groupings:
+                        type: object
+                        title: Groupings
+                        description: |-
+                          Requests the server returns results with information
```
Comment (illicitonion) [inline api/client-server/search.yaml]: Reword

File: specification/modules/search.rst [hunk 4]
```diff
@@ -40,6 +40,46 @@ The value of ``count`` may not match the number of results. For example due to
...
+
+- ``rank``, which returns the most relevant results first.
+- ``recent``, which returns the most recent results first.
+
+The default ordering is ``rank``.
+
+Groups
+------
+
+The client can request that the results are returned along with grouping
```
Comment (illicitonion) [inline specification/modules/search.rst]: What groups are allowed? What's the format?

File: specification/modules/search.rst [hunk 5]
```diff
@@ -40,6 +40,46 @@ The value of ``count`` may not match the number of results. For example due to
...
+The scope of the pagination is defined depending on where the ``next_batch``
+token was returned. For example, using a token inside a group will return more
+results from within that group.
+
+The currently supported locations for the ``next_batch`` token are:
+
+- ``search_categories.<category>.next_batch``
+- ``search_categories.<category>.groups.<group_key>.<group_id>.next_batch``
+
+A server need not support pagination, even if there are more matching results.
```
Comment (illicitonion) [inline specification/modules/search.rst]: ... in which case they should not return a next_batch token?

File: api/client-server/search.yaml [hunk 6]
```diff
@@ -107,24 +174,134 @@ paths:
...
+                          title: Room State
+                          items:
+                            type: object
+                            title: Event
+                            allOf:
+                              - "$ref": "../../event-schemas/schema/core-event-schema/room_event.json"
+                      groups:
+                        type: object
+                        title: Groups
+                        description: Any groups that were requseted.
```
Comment (illicitonion) [inline api/client-server/search.yaml]: typo requseted

File: api/client-server/search.yaml (hunk 3 above)
Comment (erikjohnston) [inline api/client-server/search.yaml]: I'm not sure if that's any better? I'm struggling a bit with explaining it :/

File: specification/modules/search.rst (hunk 5 above)
Comment (erikjohnston) [inline specification/modules/search.rst]: Done

File: specification/modules/search.rst (hunk 4 above)
Comment (erikjohnston) [inline specification/modules/search.rst]: Done

File: api/client-server/search.yaml [hunk 7]
```diff
@@ -66,8 +74,66 @@ paths:
                         type: object
                         title: Filter
                         description: |-
-                            The filter to apply to search results.
-                            This has the same format as filter API.
+                            This has the same format as v2 filter API.
```
Comment (illicitonion) [inline api/client-server/search.yaml]: v2 doesn't exist :) Please reference http://matrix.org/speculator/spec/169/client_server.html#filtering in some nice way

File: api/client-server/search.yaml [hunk 8]
```diff
@@ -107,24 +173,135 @@ paths:
...
+                                events_after:
+                                  type: array
+                                  title: Events After
+                                  description: Events just after the result.
+                                  items:
+                                    type: object
+                                    title: Event
+                                    allOf:
+                                      - "$ref": "../../event-schemas/schema/core-event-schema/room_event.yaml"
+                      state:
```
Comment (illicitonion) [inline api/client-server/search.yaml]: Add a description to... Some level of the nesting here? Right now state has no description, and its values have no concrete type beyond "object"

File: api/client-server/search.yaml (hunk 7 above)
Comment (erikjohnston) [inline api/client-server/search.yaml]: Oh, right, yes. Do you have any advice on how to reference that section nicely?

Comment (illicitonion) [inline api/client-server/search.yaml]: Nope :)

You can link to https://matrix.org/docs/spec/client_server.html#filtering explicitly maybe?

File: api/client-server/search.yaml (hunk 8 above)
Comment (erikjohnston) [inline api/client-server/search.yaml]: Done

File: api/client-server/search.yaml (hunk 7 above)
Comment (erikjohnston) [inline api/client-server/search.yaml]: I'm not sure how much I like randomly linking to paths on matrix.org, but I can't really think of a better way of doing it :/

Comment (ara4n) [inline api/client-server/search.yaml]: More importantly, I found "this has the same format as the r0 filter API" very confusing - as the filter API seems to have many different components and flavours.  When implementing this I had to grep around matrix-doc considerably before working out that I was looking for ./api/client-server/definitions/room_event_filter.json.  Or am I missing something?

File: api/client-server/search.yaml [hunk 9]
```diff
@@ -107,24 +173,139 @@ paths:
...
+                        type: object
+                        title: Current state
+                        description: |-
+                          The current state for every room in the results.
+                          This is included if the request had the the
```
Comment (illicitonion) [inline api/client-server/search.yaml]: s/the the/the/

**Comment by illicitonion** [comment]
LGTM with one comment

File: api/client-server/search.yaml (hunk 7 above)
Comment (erikjohnston) [inline api/client-server/search.yaml]: Suggestions welcome on how to do this

Comment (illicitonion) [inline api/client-server/search.yaml]: I would just say "This takes a <a href="https://matrix.org/docs/spec/%CLIENT_RELEASE_LABEL%/client_server.html#filtering">filter</a> and be happy

**Comment by erikjohnston** [comment]
I think I've addressed all the comments now? I'm not sure how the rst link in the 95255bc commit works with swagger?

**Comment by illicitonion** [comment]
LGTM

Whatever, it's probably fine, we can fix whatever we need to in the future

**Pull request merged**